src/error.rs

```diff
@@ -0,0 +1,110 @@
+use failure::{self, Context, Fail};
+use std::fmt::{self, Display};
+use std::io;
+
+/// MWixnet error definition
+#[derive(Debug)]
+pub struct Error {
+    inner: Context<ErrorKind>,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Clone, Debug, Eq, Fail, PartialEq)]
+/// MWixnet error types
+pub enum ErrorKind {
+    /// Unsupported payload version
+    #[fail(display = "Unsupported Payload Version")]
+    UnsupportedPayload,
+    /// Error from secp256k1-zkp library
+    #[fail(display = "Secp Error")]
+    SecpError,
+    /// Invalid key length for MAC initialization
+    #[fail(display = "InvalidKeyLength")]
+    InvalidKeyLength,
+    /// Wraps an io error produced when reading or writing
+    #[fail(display = "IOError")]
+    IOErr(
+        String,
+        io::ErrorKind,
+    ),
+    /// Expected a given value that wasn't found
+    #[fail(display = "UnexpectedData")]
+    UnexpectedData {
+        /// What we wanted
+        expected: Vec<u8>,
+        /// What we got
+        received: Vec<u8>,
+    },
+    /// Data wasn't in a consumable format
+    #[fail(display = "CorruptedData")]
+    CorruptedData,
+    /// Incorrect number of elements (when deserializing a vec via read_multi say).
+    #[fail(display = "CountError")]
+    CountError,
+    /// When asked to read too much data
+    #[fail(display = "TooLargeReadErr")]
+    TooLargeReadErr,
+}

```

Comment:
There should perhaps be something like a DuplicateInput and DuplicateOutput error.

Reply:
Good idea. I'll include those errors when I write the duplicate checking logic in milestone 2